clients/drcachesim/tracer/raw2trace.cpp

```diff
@@ -970,6 +1007,63 @@ instr_summary_t::construct(void *dcontext, INOUT app_pc *pc, app_pc orig_pc,
     return true;
 }
 
+const offline_entry_t *
+raw2trace_t::get_next_entry()
+{
+    if (!read_from_thread_file(tidx, &last_entry, 1))
+        return nullptr;
+    return &last_entry;
+}
+
+void
+raw2trace_t::unread_last_entry()
+{
+    unread_from_thread_file(tidx, &last_entry, 1);
+}
+
+trace_entry_t *
+raw2trace_t::get_write_buffer()
+{
+    return out_buf;
+}
+
+bool
+raw2trace_t::write(const trace_entry_t *start, const trace_entry_t *end)
+{
+    return !!out_file->write(reinterpret_cast<const char *>(start),
+                             reinterpret_cast<const char *>(end) -
+                                 reinterpret_cast<const char *>(start));
+}
+
+std::string
+raw2trace_t::write_delayed_branches(const trace_entry_t *start, const trace_entry_t *end)
+{
+    CHECK(delayed_branch[tidx].empty(), "Failed to flush delayed branch");
+    delayed_branch[tidx].insert(delayed_branch[tidx].begin(),
+                                reinterpret_cast<const char *>(start),
+                                reinterpret_cast<const char *>(end));
+    return "";
+}
+
+std::string
+raw2trace_t::on_thread_end()
+{
+    offline_entry_t entry;
+    if (read_from_thread_file(tidx, &entry, 1) || !thread_file_at_eof(tidx))
+        return "Footer is not the final entry";
+    --thread_count;
+    return "";
+}
+
+void
+raw2trace_t::log(uint level, const char *fmt, ...)
+{
+    va_list args;
+    va_start(args, fmt);
+    VPRINT(level, fmt, args);
```

Comment:
This does not look correct: this is going to pass a single param of type va_list instead of the sequence of multiple params that VPRINT expects.

Reply:
Ah, yes - thanks. Fixed.